### PR TITLE
feat: first 4 metrics tables

### LINKED ISSUE #28: Table: metric_daily_transactions
```
with
daily_txs as (
    select 
        date_trunc('day', block_timestamp) as day_date,
        count(1) as num_txs
    from harmony.prod.txs
    where day_date > current_date() - 30 -- change to increment
    group by day_date
)

select * from daily_txs
order by day_date desc
```

### LINKED ISSUE #29: Table: metric_daily_erc20_transfers
```
with
daily_transfers as (
    select 
        date_trunc('day', block_timestamp) as day_date,
        count(1) as num_transfers
    from harmony.prod.transfers
    where day_date > current_date() - 30 -- change to increment
    group by day_date
)

select * from daily_transfers
order by day_date desc
```

### LINKED ISSUE #30: Table: metric_daily_unique_users
```
with
min_address as (
    select 
        from_address,
        min(block_timestamp) as min_timestamp
    from harmony.prod.txs
    group by 1
),

final as (
    select
        date_trunc('day', min_timestamp) as min_date,
        count(from_address)
    from min_address
    group by 1
)

select * from final
order by min_date desc
```

### LINKED ISSUE #31: Table: metric_daily_gas_used
```
with
daily_gas_usage as (
    select 
        date_trunc('day', block_timestamp) as day_date,
        sum(gas_used) as gas_per_day
    from harmony.prod.blocks
    where day_date > current_date() - 30 -- change to increment
    group by day_date
)

select * from daily_gas_usage
order by day_date desc
```

### PR DESCRIPTION
**Pull request by antonyip**
Closes #28
Closes #29
Closes #30
Closes #31

### CODE REVIEW

File: macros/incremental_utils.sql [hunk 1]
```diff
@@ -6,3 +6,11 @@
...
     {% endif %}
 {%- endmacro %}
+
+{% macro incremental_last_x_days(time_col, time_in_days) -%}
+    {% if is_incremental() %}
+        {{ time_col }} >= current_date() - {{ time_in_days }}
+    {% else %}
+        true
+    {% endif %}  
+{%- endmacro %}
```
Comment (chuxinh) [inline macros/incremental_utils.sql]: is this in another PR as well? probably just merge the first one and then rebase?

File: models/metrics/metric_erc20_transfers_count.sql [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+with
+
+daily as (
+    select 
+        date_trunc('day', block_timestamp) as metric_date,
+        'daily' as metric_period,
+        count(1) as transfers_count
+    from {{ ref("transfers") }}
+    where {{ incremental_last_x_days("block_timestamp", 3) }}
+    group by 1
```
Comment (chuxinh) [inline models/metrics/metric_erc20_transfers_count.sql]:
```suggestion
    group by 1, 2
```

File: models/metrics/metric_erc20_transfers_count.sql [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+),
+
+hourly as (
+    select 
+        date_trunc('hour', block_timestamp) as metric_date,
+        'hourly' as metric_period,
+        count(1) as transfers_count
+    from {{ ref("transfers") }}
+    where {{ incremental_last_x_days("block_timestamp", 3) }}
+    group by 1
```
Comment (chuxinh) [inline models/metrics/metric_erc20_transfers_count.sql]:
```suggestion
    group by 1, 2
```

File: models/metrics/metric_erc20_transfers_count.sql [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+),
+
+minute as (
+    select 
+        date_trunc('minute', block_timestamp) as metric_date,
+        'minute' as metric_period,
+        count(1) as transfers_count
+    from {{ ref("transfers") }}
+    where {{ incremental_last_x_days("block_timestamp", 3) }}
+    group by 1
```
Comment (chuxinh) [inline models/metrics/metric_erc20_transfers_count.sql]:
```suggestion
    group by 1, 2
```

File: models/metrics/metric_erc20_transfers_count.yml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+version: 2
+
+models:
+  - name: metric_erc20_transfers_count
+    description: "The distinct number of miners within a given period"
```
Comment (chuxinh) [inline models/metrics/metric_erc20_transfers_count.yml]: should be count of erc20 transfers

File: models/metrics/metric_gas_used_count.sql [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+daily as (
+
+    select 
+        date_trunc('day', block_timestamp) as metric_date,
+        'daily' as metric_period,
+        sum(gas_used) as gas_count
+    from {{ ref("blocks") }}
+    where {{ incremental_last_x_days("block_timestamp", 3) }}
+    group by 1
```
Comment (chuxinh) [inline models/metrics/metric_gas_used_count.sql]:
```suggestion
    group by 1, 2
```

File: models/metrics/metric_gas_used_count.sql [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+hourly as (
+
+    select 
+        date_trunc('hour', block_timestamp) as metric_date,
+        'hourly' as metric_period,
+        sum(gas_used) as gas_count
+    from {{ ref("blocks") }}
+    where {{ incremental_last_x_days("block_timestamp", 3) }}
+    group by 1
```
Comment (chuxinh) [inline models/metrics/metric_gas_used_count.sql]:
```suggestion
    group by 1, 2
```

File: models/metrics/metric_gas_used_count.sql [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+minute as (
+
+    select 
+        date_trunc('minute', block_timestamp) as metric_date,
+        'minute' as metric_period,
+        sum(gas_used) as gas_count
+    from {{ ref("blocks") }}
+    where {{ incremental_last_x_days("block_timestamp", 3) }}
+    group by 1
```
Comment (chuxinh) [inline models/metrics/metric_gas_used_count.sql]:
```suggestion
    group by 1, 2
```

File: models/metrics/metric_gas_used_count.sql [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+}}
+
+with
+
+daily as (
+
+    select 
+        date_trunc('day', block_timestamp) as metric_date,
+        'daily' as metric_period,
+        sum(gas_used) as gas_count
```
Comment (chuxinh) [inline models/metrics/metric_gas_used_count.sql]: gas_total?

File: models/metrics/metric_gas_used_count.yml [hunk 10]
```diff
@@ -0,0 +1,18 @@
+version: 2
+
+models:
+  - name: metric_gas_used_count
+    description: "The distinct number of miners within a given period"
```
Comment (chuxinh) [inline models/metrics/metric_gas_used_count.yml]: update description here

File: models/metrics/metric_miner_count.sql [hunk 11]
```diff
@@ -1,26 +1,51 @@
...
 
-UNION ALL
+daily as (
+    select 
+        date_trunc('day', block_timestamp) as metric_date,
+        'daily' as metric_period,
+        count(distinct miner) as miner_count
+    from {{ ref("blocks") }}
+    where {{ incremental_last_x_days("block_timestamp", 3) }}
+    group by 1
```
Comment (chuxinh) [inline models/metrics/metric_miner_count.sql]: update the following group by

File: models/metrics/metric_transactions_count.sql [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+with
+
+daily as (
+    select 
+        date_trunc('day', block_timestamp) as metric_date,
+        'daily' as metric_period,
+        count(1) as txs_count
+    from {{ ref("txs") }}
+    where {{ incremental_last_x_days("block_timestamp", 3) }}
+    group by 1
```
Comment (chuxinh) [inline models/metrics/metric_transactions_count.sql]: update the following group by

File: models/metrics/metric_unique_users_count.sql [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    config(
+        materialized='table',
+        tags=['metrics'],
+        cluster_by=['metric_date']
+    )
+}}
+
+with
+
+min_address as (
```
Comment (chuxinh) [inline models/metrics/metric_unique_users_count.sql:11]: the logic here is only counting wallets only on the first day they are active?
not sure how useful this is. i think the block explorers, it's either looking at cumulative wallets over time, or things like daily, weekly, monthly active wallets, similar to that of monthly active user concept

**Review by chuxinh — CHANGES_REQUESTED** [review]

File: models/metrics/metric_unique_users_count.sql (hunk 13 above)
Comment (antonyip) [reply models/metrics/metric_unique_users_count.sql:11]: You can do that on the frontend.. but I guess we can add a cumulative column.

File: models/metrics/metric_unique_users_count.sql [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+),
+
+daily as (
+
+    select
+        date_trunc('day', min_timestamp) as metric_date,
+        'daily' as metric_period,
+        count(from_address) as unique_users_count,
+        sum(count(from_address)) over (order by metric_date) as unique_users_cumulative
```
Comment (chuxinh) [inline models/metrics/metric_unique_users_count.sql]: you can directly use `unique_users_count` from above
```suggestion
        sum(unique_users_count) over (partition by metric_period order by metric_date) as unique_users_cumulative
```

File: models/metrics/metric_unique_users_count.sql [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+),
+
+hourly as (
+
+    select
+        date_trunc('hour', min_timestamp) as metric_date,
+        'hourly' as metric_period,
+        count(from_address) as unique_users_count,
+        sum(count(from_address)) over (order by metric_date) as unique_users_cumulative
```
Comment (chuxinh) [inline models/metrics/metric_unique_users_count.sql]: same as above

File: models/metrics/metric_unique_users_count.sql [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+),
+
+minute as (
+
+    select
+        date_trunc('minute', min_timestamp) as metric_date,
+        'minute' as metric_period,
+        count(from_address) as unique_users_count,
+        sum(count(from_address)) over (order by metric_date) as unique_users_cumulative
```
Comment (chuxinh) [inline models/metrics/metric_unique_users_count.sql]: same as above

**Review by chuxinh — APPROVED** [review]
LGTM after nits

**Pull request merged**